src/v/iceberg/merge_append_action.cc

```diff
@@ -0,0 +1,552 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#include "iceberg/merge_append_action.h"
+
+#include "base/units.h"
+#include "base/vlog.h"
+#include "iceberg/logger.h"
+#include "iceberg/manifest.h"
+#include "iceberg/manifest_file_packer.h"
+#include "iceberg/manifest_list.h"
+#include "iceberg/snapshot.h"
+#include "iceberg/table_requirement.h"
+#include "iceberg/values_bytes.h"
+#include "random/generators.h"
+
+#include <iterator>
+#include <limits>
+
+namespace iceberg {
+
+namespace {
+manifest_path get_manifest_path(
+  const ss::sstring& location, const uuid_t& commit_uuid, size_t num) {
+    return manifest_path{
+      fmt::format("{}/metadata/{}-m{}.avro", location, commit_uuid, num)};
+}
+manifest_list_path get_manifest_list_path(
+  const ss::sstring& location,
+  snapshot_id snap_id,
+  const uuid_t& commit_uuid,
+  size_t num) {
+    return manifest_path{fmt::format(
+      "{}/metadata/snap-{}-{}-{}.avro", location, snap_id(), commit_uuid, num)};
+}
+
+action::errc to_action_errc(manifest_io::errc e) {
+    switch (e) {
+    case manifest_io::errc::failed:
+        return action::errc::io_failed;
+    case manifest_io::errc::shutting_down:
+        return action::errc::shutting_down;
+    case manifest_io::errc::timedout:
+        // NOTE: treat IO timeouts the same as other IO failures.
+        // TODO: build out retry logic.
+        return action::errc::io_failed;
+    }
+}
+
+snapshot_id random_snap_id() {
+    return snapshot_id{random_generators::get_int<int64_t>(
+      0, std::numeric_limits<int64_t>::max())};
```

Comment:
is the according to iceberg spec, or we decided on this? what properties does it need?

Reply:
The spec is quite mum about what this should be. Implementations appear to be random. Here's py-iceberg's that pulls a random int64 out of a uuid:

https://github.com/apache/iceberg-python/blob/e5a58b34dd830c6ffea11649613b693f70f7cbb4/pyiceberg/table/metadata.py#L322-L333

Reply:
does it need to be as strong as a uuid? our random source is pretty much just the default c++ standard library prng

Reply:
I don't think so, IIUC it just needs to be unique within a given table to avoid collisions between multiple writers (though in our case, we only expect a single writer at a time per table)